File: wear/app/src/main/java/app/imuatrak/wear/services/Aggregator.kt
package app.imuatrak.wear.services

import app.imuatrak.wear.models.*
import kotlin.math.*

// Mirrors src/services/aggregator.ts — same logic, same results.

object Aggregator {

    fun totals(track: List<WatchTrackPoint>, strokeCount: Int): WatchTotals {
        if (track.size < 2) return WatchTotals.empty()
        var distM = 0.0; var movingDur = 0.0; var maxSpeed = 0.0; var elevGain = 0.0
        for (i in 1 until track.size) {
            val prev = track[i-1]; val cur = track[i]
            val d = haversine(prev.lat, prev.lon, cur.lat, cur.lon)
            distM += d
            val dt = cur.t - prev.t
            if (cur.speedMps > 0.5 || d > 0.5) movingDur += dt
            maxSpeed = max(maxSpeed, cur.speedMps)
            val dAlt = cur.altM - prev.altM
            if (dAlt > 0) elevGain += dAlt
        }
        val dur = track.last().t - track.first().t
        val avgSpeed = if (dur > 0) distM / dur else 0.0
        val avgPace = if (distM > 0) dur / (distM / 1000) else 0.0
        val avgSr = if (dur > 0) strokeCount / (dur / 60) else 0.0
        return WatchTotals(distM, dur, movingDur, avgPace, avgSpeed, maxSpeed, strokeCount, avgSr, 0.0, elevGain)
    }

    fun hrSummary(track: List<WatchTrackPoint>): WatchHrSummary {
        val hrs = track.mapNotNull { it.hr }
        if (hrs.isEmpty()) return WatchHrSummary.empty()
        val zoneBounds = listOf(0, 120, 140, 160, 175, 220)
        val zoneTimes = DoubleArray(5)
        for (i in 1 until track.size) {
            val hr = track[i].hr ?: continue
            val dt = track[i].t - track[i-1].t
            var z = 0
            for (k in zoneBounds.indices) { if (hr >= zoneBounds[k]) z = k }
            zoneTimes[z.coerceIn(0, 4)] += dt
        }
        val avg = hrs.average().toInt()
        val maxHr = hrs.max()
        val bounds = listOf(0 to 120, 120 to 140, 140 to 160, 160 to 175, 175 to 220)
        val zones = bounds.mapIndexed { i, (mn, mx) -> WatchHrZone(i, mn, mx, zoneTimes[i]) }
        return WatchHrSummary(avg, maxHr, zones)
    }

    fun splits(track: List<WatchTrackPoint>): List<WatchSplit> {
        val unitM = 1000.0
        val result = mutableListOf<WatchSplit>()
        var splitStart = 0; var accumulated = 0.0; var index = 0
        val hrs = mutableListOf<Double>(); val srs = mutableListOf<Double>()
        for (i in 1 until track.size) {
            accumulated += haversine(track[i-1].lat, track[i-1].lon, track[i].lat, track[i].lon)
            track[i].hr?.let { hrs.add(it.toDouble()) }
            track[i].strokeRate?.let { srs.add(it) }
            if (accumulated >= unitM) {
                val dur = track[i].t - track[splitStart].t
                result.add(WatchSplit(index++, accumulated, dur,
                    if (hrs.isEmpty()) 0.0 else hrs.average(),
                    if (srs.isEmpty()) 0.0 else srs.average(),
                    if (dur > 0) accumulated / dur else 0.0))
                splitStart = i; accumulated = 0.0; hrs.clear(); srs.clear()
            }
        }
        return result
    }

    fun downsample(track: List<WatchTrackPoint>, maxPoints: Int = 200): List<WatchTrackPoint> {
        if (track.size <= maxPoints) return track
        val step = track.size.toDouble() / maxPoints
        return (0 until maxPoints).map { track[(it * step).toInt()] }
    }

    private fun haversine(lat1: Double, lon1: Double, lat2: Double, lon2: Double): Double {
        val R = 6371000.0
        val φ1 = Math.toRadians(lat1); val φ2 = Math.toRadians(lat2)
        val Δφ = Math.toRadians(lat2 - lat1); val Δλ = Math.toRadians(lon2 - lon1)
        val a = sin(Δφ/2).pow(2) + cos(φ1)*cos(φ2)*sin(Δλ/2).pow(2)
        return R * 2 * atan2(sqrt(a), sqrt(1-a))
    }
}

File: apple-watch/Sources/Services/Aggregator.swift
import Foundation
import CoreLocation

// ── Aggregator.swift ──────────────────────────────────────────────────────────
// Computes WatchTotals, WatchHrSummary, and WatchSplit[] from a raw track.
// Logic mirrors src/services/aggregator.ts exactly.

enum Aggregator {

    static func totals(_ track: [WatchTrackPoint], strokeCount: Int) -> WatchTotals {
        guard track.count >= 2 else { return .empty() }

        var distanceM = 0.0
        var movingDurationSec = 0.0
        var maxSpeedMps = 0.0
        var elevGain = 0.0

        for i in 1..<track.count {
            let prev = track[i-1], cur = track[i]
            let d = haversine(lat1: prev.lat, lon1: prev.lon, lat2: cur.lat, lon2: cur.lon)
            distanceM += d
            let dt = cur.t - prev.t
            if cur.speedMps > 0.5 || d > 0.5 { movingDurationSec += dt }
            maxSpeedMps = max(maxSpeedMps, cur.speedMps)
            let dAlt = cur.altM - prev.altM
            if dAlt > 0 { elevGain += dAlt }
        }

        let durationSec = track.last!.t - track.first!.t
        let avgSpeedMps = durationSec > 0 ? distanceM / durationSec : 0
        let avgPace = distanceM > 0 ? (durationSec / (distanceM / 1000)) : 0
        let avgStrokeRate = durationSec > 0 ? Double(strokeCount) / (durationSec / 60) : 0

        return WatchTotals(
            distanceMeters: distanceM,
            durationSec: durationSec,
            movingDurationSec: movingDurationSec,
            avgPaceSecPerKm: avgPace,
            avgSpeedMps: avgSpeedMps,
            maxSpeedMps: maxSpeedMps,
            strokeCount: strokeCount,
            avgStrokeRate: avgStrokeRate,
            calories: 0,
            elevationGainM: elevGain
        )
    }

    static func hrSummary(_ track: [WatchTrackPoint]) -> WatchHrSummary {
        let hrs = track.compactMap { $0.hr }
        guard !hrs.isEmpty else { return .empty() }

        let zoneBounds = [0, 120, 140, 160, 175, 220]
        var zoneTimes = [Double](repeating: 0, count: 5)
        for i in 1..<track.count {
            guard let hr = track[i].hr else { continue }
            let dt = track[i].t - track[i-1].t
            var z = 0
            for k in 0..<zoneBounds.count { if hr >= zoneBounds[k] { z = k } }
            z = max(0, min(4, z))
            zoneTimes[z] += dt
        }

        let avg = Int(hrs.reduce(0, +) / hrs.count)
        let maxHr = hrs.max() ?? 0
        let bounds = [(0,120),(120,140),(140,160),(160,175),(175,220)]
        let zones = bounds.enumerated().map { i, b in
            WatchHrZone(zone: i, minBpm: b.0, maxBpm: b.1, timeSec: zoneTimes[i])
        }
        return WatchHrSummary(avg: avg, max: maxHr, zones: zones)
    }

    static func splits(_ track: [WatchTrackPoint], imperial: Bool = false) -> [WatchSplit] {
        let unitM: Double = imperial ? 1609.344 : 1000.0
        var splits: [WatchSplit] = []
        var splitStart = 0
        var accumulated = 0.0
        var splitHrs: [Double] = []
        var splitStrokes: [Double] = []
        var index = 0

        for i in 1..<track.count {
            let d = haversine(lat1: track[i-1].lat, lon1: track[i-1].lon,
                              lat2: track[i].lat, lon2: track[i].lon)
            accumulated += d
            if let hr = track[i].hr { splitHrs.append(Double(hr)) }
            if let sr = track[i].strokeRate { splitStrokes.append(sr) }

            if accumulated >= unitM {
                let dur = track[i].t - track[splitStart].t
                let avgHr = splitHrs.isEmpty ? 0.0 : splitHrs.reduce(0,+)/Double(splitHrs.count)
                let avgSr = splitStrokes.isEmpty ? 0.0 : splitStrokes.reduce(0,+)/Double(splitStrokes.count)
                let avgSpd = dur > 0 ? accumulated / dur : 0
                splits.append(WatchSplit(index: index, distanceM: accumulated,
                                        durationSec: dur, avgHr: avgHr,
                                        avgStrokeRate: avgSr, avgSpeedMps: avgSpd))
                index += 1
                splitStart = i
                accumulated = 0
                splitHrs = []
                splitStrokes = []
            }
        }
        return splits
    }

    /// Ramer-Douglas-Peucker downsampling to ≤maxPoints track summary points.
    static func downsample(_ track: [WatchTrackPoint], maxPoints: Int = 200) -> [WatchTrackPoint] {
        guard track.count > maxPoints else { return track }
        let step = Double(track.count) / Double(maxPoints)
        return (0..<maxPoints).map { i in track[Int(Double(i) * step)] }
    }

    // Haversine distance in metres between two WGS-84 coordinates
    private static func haversine(lat1: Double, lon1: Double, lat2: Double, lon2: Double) -> Double {
        let R = 6371000.0
        let φ1 = lat1 * .pi / 180, φ2 = lat2 * .pi / 180
        let Δφ = (lat2 - lat1) * .pi / 180
        let Δλ = (lon2 - lon1) * .pi / 180
        let a = sin(Δφ/2)*sin(Δφ/2) + cos(φ1)*cos(φ2)*sin(Δλ/2)*sin(Δλ/2)
        return R * 2 * atan2(a.squareRoot(), (1-a).squareRoot())
    }
}
